### PR TITLE
[Fluid] Explicit compressible Navier-Stokes element

### PR DESCRIPTION
**Pull request by rubenzorrilla**
This add the explicit version of the current implicit compressible Navier-Stokes element.

It is generated by using the automatic differentiation in such a way that the Gauss pt. loop is avoided. The explicit RHS contributions are assembled in the reaction variables to be compatible with the explicit DOF based strategy we are working on.

Note that this version doesn't include the shock capturing technique as we still need to discuss about it.

The explicit residuals are checked with the existing implicit element ones (without shock capturing).

### CODE REVIEW

File: applications/FluidDynamicsApplication/symbolic_generation/compressible_navier_stokes/compressible_navier_stokes_explicit_cpp_template_with_integration.cpp [hunk 1]
```diff
@@ -0,0 +1,603 @@
...
+    const double &U_2_1 = data.U(2, 1);
+    const double &U_2_2 = data.U(2, 2);
+    const double &U_2_3 = data.U(2, 3);
+
+    // Hardcoded shape functions for linear triangular element
+    // This is explicitly done to minimize the allocation and matrix acceses
+    // The notation N_i_j means shape function for node j in Gauss pt. i
+    const double one_sixt = 1.0/6.0;
+    const double two_third = 2.0/3.0;
+    const double N_0_0 = one_sixt;
```
Comment (RiccardoRossi) [inline applications/FluidDynamicsApplication/symbolic_generation/compressible_navier_stokes/compressible_navier_stokes_explicit_cpp_template_with_integration.cpp]: in a future PR i would put this in the symbolic so to enjoy simplifications.

File: applications/FluidDynamicsApplication/symbolic_generation/compressible_navier_stokes/compressible_navier_stokes_explicit_cpp_template_with_integration.cpp [hunk 2]
```diff
@@ -0,0 +1,603 @@
...
+    const double &U_3_0 = data.U(3, 0);
+    const double &U_3_1 = data.U(3, 1);
+    const double &U_3_2 = data.U(3, 2);
+    const double &U_3_3 = data.U(3, 3);
+    const double &U_3_4 = data.U(3, 4);
+
+    // Hardcoded shape functions for linear tetrahedra element
+    // This is explicitly done to minimize the alocation and matrix acceses
+    // The notation N_i_j means shape function for node j in Gauss pt. i
+    const double N_0_0 = 0.58541020;
```
Comment (RiccardoRossi) [inline applications/FluidDynamicsApplication/symbolic_generation/compressible_navier_stokes/compressible_navier_stokes_explicit_cpp_template_with_integration.cpp]: same as above. in a future PR this should go in the symbolic gen

File: applications/FluidDynamicsApplication/symbolic_generation/compressible_navier_stokes/compressible_navier_stokes_explicit_cpp_template_with_integration.cpp [hunk 3]
```diff
@@ -0,0 +1,603 @@
...
+
+template <>
+void CompressibleNavierStokesExplicit<2>::AddExplicitContribution(const ProcessInfo &rCurrentProcessInfo)
+{
+    constexpr IndexType dim = 2;
+    constexpr IndexType n_nodes = 3;
+    constexpr IndexType block_size = 4;
+
+    // Calculate the explicit residual vector
+    VectorType rhs;
```
Comment (RiccardoRossi) [inline applications/FluidDynamicsApplication/symbolic_generation/compressible_navier_stokes/compressible_navier_stokes_explicit_cpp_template_with_integration.cpp]: I'd create an auxiliary method so the RHS can use bounded matrices.

File: applications/FluidDynamicsApplication/symbolic_generation/compressible_navier_stokes/compressible_navier_stokes_explicit_cpp_template_with_integration.cpp [hunk 4]
```diff
@@ -0,0 +1,603 @@
...
+    MatrixType &rMassMatrix,
+    const ProcessInfo &rCurrentProcessInfo)
+{
+    constexpr IndexType n_nodes = 4;
+    constexpr IndexType block_size = 5;
+
+    // Hardcoded shape functions for linear tetrahedra element
+    // This is explicitly done to minimize the alocation and matrix acceses
+    // The notation N_i_j means shape function for node j in Gauss pt. i
+    const double N_0_0 = 0.58541020;
```
Comment (RiccardoRossi) [inline applications/FluidDynamicsApplication/symbolic_generation/compressible_navier_stokes/compressible_navier_stokes_explicit_cpp_template_with_integration.cpp]: take care about this. it should be Ni*Nj (integrated)

File: applications/FluidDynamicsApplication/symbolic_generation/compressible_navier_stokes/generate_compressible_navier_stokes_explicit_with_integration.py [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+import pprint
+
+from params_dict import params
+import generate_convective_flux
+import generate_diffusive_flux
+import generate_source_term
+import generate_stabilization_matrix
+
+# dim = params["dim"]         # Define Dimension in params.py
+do_simplifications = False
```
Comment (RiccardoRossi) [inline applications/FluidDynamicsApplication/symbolic_generation/compressible_navier_stokes/generate_compressible_navier_stokes_explicit_with_integration.py:64]: in a future PR, when it works correctly, we should have a True here

File: applications/FluidDynamicsApplication/symbolic_generation/compressible_navier_stokes/generate_compressible_navier_stokes_explicit_with_integration.py [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+    # Shape functions and Gauss pts. settings
+    if(dim == 2):
+        n_nodes = 3
+        n_gauss = 3
+    elif(dim == 3):
+        n_nodes = 4
+        n_gauss = 4
+
+    DN = DefineMatrix('DN', n_nodes, dim)
+    mat_N = DefineMatrix('mat_N', n_gauss, n_nodes)
```
Comment (RiccardoRossi) [inline applications/FluidDynamicsApplication/symbolic_generation/compressible_navier_stokes/generate_compressible_navier_stokes_explicit_with_integration.py]: numeric values of shape functions should be given here

File: applications/FluidDynamicsApplication/symbolic_generation/compressible_navier_stokes/generate_compressible_navier_stokes_explicit_with_integration.py [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+    for i_gauss in range(n_gauss):
+        print("\tGauss point: " + str(i_gauss))
+        rv_gauss = rv.copy()
+        res_m_gauss = res_m.copy()
+        res_e_gauss = res_e.copy()
+
+        ## Get Gauss point geometry data
+        N = DefineVector('N', n_nodes)
+        for i in range(n_nodes):
+            N[i] = mat_N[i_gauss, i]
```
Comment (RiccardoRossi) [inline applications/FluidDynamicsApplication/symbolic_generation/compressible_navier_stokes/generate_compressible_navier_stokes_explicit_with_integration.py:237]: give numeric values to this

File: applications/FluidDynamicsApplication/symbolic_generation/compressible_navier_stokes/compressible_navier_stokes_explicit_cpp_template_with_integration.cpp (hunk 4 above)
Comment (rubenzorrilla) [reply applications/FluidDynamicsApplication/symbolic_generation/compressible_navier_stokes/compressible_navier_stokes_explicit_cpp_template_with_integration.cpp]: Copy paste error. I'd put the analytical values.

File: applications/FluidDynamicsApplication/symbolic_generation/compressible_navier_stokes/generate_compressible_navier_stokes_explicit_with_integration.py (hunk 6 above)
Comment (rubenzorrilla) [reply applications/FluidDynamicsApplication/symbolic_generation/compressible_navier_stokes/generate_compressible_navier_stokes_explicit_with_integration.py]: Done.

File: applications/FluidDynamicsApplication/symbolic_generation/compressible_navier_stokes/compressible_navier_stokes_explicit_cpp_template_with_integration.cpp (hunk 3 above)
Comment (rubenzorrilla) [reply applications/FluidDynamicsApplication/symbolic_generation/compressible_navier_stokes/compressible_navier_stokes_explicit_cpp_template_with_integration.cpp]: Done.

File: applications/FluidDynamicsApplication/symbolic_generation/compressible_navier_stokes/compressible_navier_stokes_explicit_cpp_template_with_integration.cpp (hunk 2 above)
Comment (rubenzorrilla) [reply applications/FluidDynamicsApplication/symbolic_generation/compressible_navier_stokes/compressible_navier_stokes_explicit_cpp_template_with_integration.cpp]: Done.

File: applications/FluidDynamicsApplication/symbolic_generation/compressible_navier_stokes/compressible_navier_stokes_explicit_cpp_template_with_integration.cpp (hunk 1 above)
Comment (rubenzorrilla) [reply applications/FluidDynamicsApplication/symbolic_generation/compressible_navier_stokes/compressible_navier_stokes_explicit_cpp_template_with_integration.cpp]: Done.

File: applications/FluidDynamicsApplication/symbolic_generation/compressible_navier_stokes/generate_compressible_navier_stokes_explicit_with_integration.py (hunk 7 above)
Comment (rubenzorrilla) [reply applications/FluidDynamicsApplication/symbolic_generation/compressible_navier_stokes/generate_compressible_navier_stokes_explicit_with_integration.py:237]: Done.

File: applications/FluidDynamicsApplication/symbolic_generation/compressible_navier_stokes/generate_compressible_navier_stokes_explicit_with_integration.py (hunk 5 above)
Comment (rubenzorrilla) [reply applications/FluidDynamicsApplication/symbolic_generation/compressible_navier_stokes/generate_compressible_navier_stokes_explicit_with_integration.py:64]: I switched it to `True` and after one night trying to generate the element it didn't finished yet. I'd leave it as `False`. In the final steps of the implementation we can switch it on and generate the final version of the cpp.

**Comment by rubenzorrilla** [comment]
@RiccardoRossi I've implemented your efficiency suggestions. However the inertial contribution of the subscales is still missing until we discuss it.

File: applications/FluidDynamicsApplication/symbolic_generation/compressible_navier_stokes/generate_compressible_navier_stokes_explicit_with_integration.py (hunk 5 above)
Comment (RiccardoRossi) [reply applications/FluidDynamicsApplication/symbolic_generation/compressible_navier_stokes/generate_compressible_navier_stokes_explicit_with_integration.py:64]: ok

**Review by RiccardoRossi — APPROVED** [review]

**Pull request merged**